util/llbutil/authprovider/cloudstoredauthprovider/authprovider.go

```diff
@@ -0,0 +1,371 @@
+func (ap *authProvider) getProjectOrUserSecret(ctx context.Context, org, project, secretName string) (string, error) {
+	if org == "" && project == "" {
+		secret, err := ap.cloudClient.GetUserSecret(ctx, secretName)
+		if err != nil {
+			if errors.Is(err, secrets.ErrNotFound) {
+				return "", nil // treat non-exists and empty secrets the same way
+			}
+			return "", fmt.Errorf("failed to lookup user secret %s: %w", secretName, err)
+		}
+		return secret.Value, nil
+	}
+
+	secret, err := ap.cloudClient.GetProjectSecret(ctx, org, project, secretName)
+	if err != nil {
+		if errors.Is(err, secrets.ErrNotFound) {
+			return "", nil // treat non-exists and empty secrets the same way
+		}
+		return "", fmt.Errorf("failed to lookup project secret %s/%s/%s: %w", org, project, secretName, err)
+	}
+	return secret.Value, nil
+}
+
+func (ap *authProvider) getAuthConfigForProject(ctx context.Context, org, project, host string) (*authConfig, error) {
+	pathPrefix := fmt.Sprintf("std/registry/%s/", host)
+	var fullPathPrefix string // only used for error messages
+	if org == "" && project == "" {
+		fullPathPrefix = fmt.Sprintf("/user/%s", pathPrefix)
+	} else {
+		fullPathPrefix = fmt.Sprintf("/%s/%s/%s", org, project, pathPrefix)
+	}
+	userPath := pathPrefix + "username"
+	passwordPath := pathPrefix + "password"
+
+	ap.console.VerbosePrintf("looking up %susername", fullPathPrefix)
+	username, err := ap.getProjectOrUserSecret(ctx, org, project, userPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ap.console.VerbosePrintf("looking up %spassword", fullPathPrefix)
+	password, err := ap.getProjectOrUserSecret(ctx, org, project, passwordPath)
+	if err != nil {
+		return nil, err
+	}
+	// TODO look for insecure and http config options (not sure how these options will be propigated to the rest of buildkit)
+
+	if username == "" && password != "" {
+		return nil, fmt.Errorf("found %s/username, but no %s/password", fullPathPrefix, fullPathPrefix)
+	}
+	if username != "" && password == "" {
+		return nil, fmt.Errorf("found %s/password, but no %s/username", fullPathPrefix, fullPathPrefix)
+	}
+
+	if username == "" && password == "" {
+		return nil, ErrNoCredentialsFound
+	}
+
+	return &authConfig{
+		ac: &types.AuthConfig{
+			ServerAddress: host,
+			Username:      username,
+			Password:      password,
+		},
+		loc: fullPathPrefix,
+	}, nil
+}
+
+// getAuthConfig was re-written to make use of earthly-cloud based credentials
+func (ap *authProvider) getAuthConfig(ctx context.Context, host string) (*authConfig, error) {
+	ap.mu.Lock()
+	defer ap.mu.Unlock()
+
+	// check the cache
+	ac, ok := ap.authConfigCache[host]
+	if ok {
+		return ac, nil
+	}
+
+	// check unser user's secrets
+	ac, err := ap.getAuthConfigForProject(ctx, "", "", host)
+	if err == nil {
+		ap.authConfigCache[host] = ac
+		return ac, nil
+	} else if err != ErrNoCredentialsFound {
```

Comment:
`else` is unnecessary here - the `if` ends in a `return`
```suggestion
	}
	if err != ErrNoCredentialsFound {
```